### PR TITLE
[completion] [unicode] Provide unicode completion server-side

### LINKED ISSUE #189: Unicode input
It would be good to have a mechanism like CoqIDE has where you can press `\` and then type in a unicode character name. It is possible to use extensions to insert unicode characters, but they don't appear in the completion interface.

### PR DESCRIPTION
**Pull request by ejgallego**
Works fine modulo client bugs; we provide a new config option for two
different completion sets.

Fixes #189

Co-authored-by: Ali Caglayan <alizter@gmail.com>
Co-authored-by: Ramkumar Ramachandra <r@artagnon.com>

### CODE REVIEW

**Comment by ejgallego** [comment]
@Alizter can you test this now? Feel free to edit for more symbols, etc...

**Comment by Alizter** [comment]
I am unable to trigger the completion when typing. Here are 3 cases I tried:
- type `\`, nothing comes up, press Ctrl + Space then the completion menu comes up with the correct items.
- type `\G`, nothing comes up, press Ctrl + Space still nothing
- press Ctrl+Space, menu comes up, press \, no items.

The trigger seems to be wrong, and the menu doesn't work if partially filled in. You need to press Ctrl + Space when only on the backspace.

**Comment by ejgallego** [comment]
> type \, nothing comes up, press Ctrl + Space then the completion menu comes up with the correct items.

That's the only case we support, strange it doesn't trigger for you (works fine here).

What VSCode version are you on?

What does the init request response say about completion?

**Comment by ejgallego** [comment]
IMHO makes sense to only trigger unicode completion on `\` , in most others contexts we should serve regular tactic / etc... stuff.

**Comment by ejgallego** [comment]
Oh likely what you are seeing is your request stolen, what does the log say about completion?

Is a request submitted when you press `\` but later cancelled?

**Comment by ejgallego** [comment]
@Alizter how does it work now?

**Comment by Alizter** [comment]
Still the same. Just typing `\` does not trigger the completion. I can only get the completion to trigger manually with Ctrl + Space.

File: fleche/debug.ml [hunk 1]
```diff
@@ -19,7 +19,7 @@ let parsing = false || all
 let scan = false || all
 
 (* Backtraces *)
-let backtraces = false || all
```
Comment (Alizter) [inline fleche/debug.ml:22]: Undo pls

**Comment by ejgallego** [comment]
> I get the following when I press \:

Umm, thanks, but I need the trace from the moment before the `didChange` notification in response to your keypress is triggered, can you get that?

**Comment by Alizter** [comment]
Here is a more full log:
```
Trace - 01:33:55] [init]: custom client options:
[Trace - 01:33:55] [init]: [init]: {
          "client_version": "0.1.4",
          "eager_diagnostics": true,
          "ok_diagnostics": false,
          "goal_after_tactic": false,
          "show_coq_info_messages": false,
          "show_notices_as_diagnostics": false,
          "admit_on_bad_qed": true
        }
[Trace - 01:33:55] [client_version]: 0.1.4
[Trace - 01:33:55] Received response 'initialize - (0)' in 225ms.
[Trace - 01:33:55] Sending notification 'initialized'.
[Trace - 01:33:55] Received notification 'window/logMessage'.
[Info  - 01:33:55] Server initialized
[Trace - 01:33:55] Sending notification 'textDocument/didOpen'.
[Trace - 01:33:55] Sending request 'textDocument/documentSymbol - (1)'.
[Trace - 01:33:55] [workspace]: initialized
[Trace - 01:33:55] Received notification 'window/logMessage'.
[Info  - 01:33:55] Configuration loaded from Command-line arguments
 - coqlib is at: /home/ali/repos/coq-lsp/_build/install/default//lib/coq
 - Modules [Coq.Init.Prelude] will be loaded by default
 - 2 Coq path directory bindings in scope; 22 Coq plugin directory bindings in scope
[Trace - 01:33:55] [process_queue]: Serving Request: initialized
[Trace - 01:33:55] [process_queue]: Serving Request: textDocument/didOpen
[Trace - 01:33:56] [process_queue]: Serving Request: textDocument/documentSymbol
[Trace - 01:33:56] [request]: postponing rq : 1
[Trace - 01:33:56] [process_queue]: resuming document checking
[Trace - 01:33:56] [check]: resuming [v: 39], from: 0 l: 0
[Trace - 01:33:56] Received notification '$/coq/fileProgress'.
[Trace - 01:33:56] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] [recovery]: ({ l: 55, c: 0 | o: 940 }--{ l: 55, c: 30 | o: 970 }) stm: 547262707 | st 689204304
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] [recovery]: ({ l: 68, c: 0 | o: 1106 }--{ l: 68, c: 18 | o: 1124 }) stm: 547262707 | st 689204304
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] [recovery]: ({ l: 78, c: 0 | o: 1262 }--{ l: 78, c: 22 | o: 1284 }) stm: 547262707 | st 689204304
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] Received notification '$/coq/fileProgress'.
[Trace - 01:33:57] [check]: done [1674520437.47]: document fully checked with pos ({ l: 106, c: 0 | o: 1676 }--
{ l: 106, c: 0 | o: 1677 })
[Trace - 01:33:57] [cache]: hashing: 0.000908 | parsing: 0.017746 | exec: 1.115927
[Trace - 01:33:57] [cache]: cache hit rate: 0.00
[Trace - 01:33:57] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:33:57] [[wake up]]: rq: 1 | {k:doc}
[Trace - 01:33:57] Received response 'textDocument/documentSymbol - (1)' in 1945ms.
[Trace - 01:34:03] Sending notification 'textDocument/didSave'.
[Trace - 01:34:03] [process_queue]: Serving Request: textDocument/didSave
[Trace - 01:34:05] Sending request 'textDocument/hover - (2)'.
[Trace - 01:34:05] Sending notification '$/cancelRequest'.
[Trace - 01:34:05] [process_queue]: Serving Request: textDocument/hover
[Trace - 01:34:05] Received response 'textDocument/hover - (2)' in 62ms.
[Trace - 01:34:05] [process_queue]: Serving Request: $/cancelRequest
[Trace - 01:34:07] Sending request 'proof/goals - (3)'.
[Trace - 01:34:07] [process_queue]: Serving Request: proof/goals
[Trace - 01:34:07] Received response 'proof/goals - (3)' in 25ms.
[Trace - 01:34:11] Sending request 'proof/goals - (4)'.
[Trace - 01:34:11] [process_queue]: Serving Request: proof/goals
[Trace - 01:34:11] Received response 'proof/goals - (4)' in 92ms.
[Trace - 01:34:15] Sending request 'textDocument/hover - (5)'.
[Trace - 01:34:15] [process_queue]: Serving Request: textDocument/hover
[Trace - 01:34:15] Received response 'textDocument/hover - (5)' in 93ms.
[Trace - 01:34:17] Sending request 'textDocument/completion - (6)'.
[Trace - 01:34:17] Sending notification 'textDocument/didChange'.
[Trace - 01:34:17] Sending request 'proof/goals - (7)'.
[Trace - 01:34:17] [process_queue]: Serving Request: textDocument/completion
[Trace - 01:34:17] [completion]: start
[Trace - 01:34:17] [completion]: line: 
[Trace - 01:34:17] [completion]: index: 0
[Trace - 01:34:17] [process_queue]: bt=true
[Trace - 01:34:17] [print_exn [OCaml]]: Invalid_argument("index out of bounds")
[Trace - 01:34:17] [print_exn [Coq  ]]: Anomaly "Uncaught exception Invalid_argument("index out of bounds")."
Please report at http://coq.inria.fr/bugs/.
Raised by primitive operation at Dune__exe__Rq_completion.get_char_at_point in file "controller/rq_completion.ml", line 72, characters 9-32
Called from Dune__exe__Rq_completion.completion in file "controller/rq_completion.ml", line 79, characters 8-37
Called from Dune__exe__Coq_lsp.do_position_request in file "controller/coq_lsp.ml", line 249, characters 26-47
Called from Dune__exe__Coq_lsp.dispatch_request in file "controller/coq_lsp.ml" (inlined), line 394, characters 2-35
Called from Dune__exe__Coq_lsp.dispatch_request in file "controller/coq_lsp.ml", line 397, characters 6-48
Called from Dune__exe__Coq_lsp.dispatch_or_resume_check in file "controller/coq_lsp.ml", line 440, characters 6-43

[Trace - 01:34:17] [print_bt  [OCaml]]: Raised by primitive operation at Dune__exe__Rq_completion.get_char_at_point in file "controller/rq_completion.ml", line 72, characters 9-32
Called from Dune__exe__Rq_completion.completion in file "controller/rq_completion.ml", line 79, characters 8-37
Called from Dune__exe__Coq_lsp.do_position_request in file "controller/coq_lsp.ml", line 249, characters 26-47
Called from Dune__exe__Coq_lsp.dispatch_request in file "controller/coq_lsp.ml" (inlined), line 394, characters 2-35
Called from Dune__exe__Coq_lsp.dispatch_request in file "controller/coq_lsp.ml", line 397, characters 6-48
Called from Dune__exe__Coq_lsp.dispatch_or_resume_check in file "controller/coq_lsp.ml", line 440, characters 6-43

[Trace - 01:34:17] [process_queue]: Serving Request: textDocument/didChange
[Trace - 01:34:17] [bump file]: file:///home/ali/repos/coq-lsp/examples/ex0.v / version: 40
[Trace - 01:34:17] [prefix]: common prefix offset found at 1104
[Trace - 01:34:17] [prefix]: resuming from ({ l: 64, c: 0 | o: 1079 }--{ l: 64, c: 23 | o: 1102 })
[Trace - 01:34:17] [bump file took]: 0.000085
[Trace - 01:34:17] [process_queue]: Serving Request: proof/goals
[Trace - 01:34:17] [request]: postponing rq : 7
[Trace - 01:34:17] [process_queue]: resuming document checking
[Trace - 01:34:17] [check]: resuming [v: 40], from: 1102 l: 64
[Trace - 01:34:17] [resume]: last node :({ l: 64, c: 0 | o: 1079 }--{ l: 64, c: 23 | o: 1102 })
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] [beyond_target]: target reached ({ l: 66, c: 1 | o: 1124 }--{ l: 66, c: 1 | o: 1125 })
[Trace - 01:34:17] [beyond_target]: target is {cutpoint l: 66 | c: 01
[Trace - 01:34:17] [check]: done [1674520457.40]: document stopped with pos ({ l: 66, c: 1 | o: 1124 }--
{ l: 66, c: 1 | o: 1125 })
[Trace - 01:34:17] [cache]: hashing: 0.000908 | parsing: 0.017962 | exec: 1.115927
[Trace - 01:34:17] [cache]: no stats
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] [[wake up]]: rq: 7 | {k:pos | l: 66, c: 1}
[Trace - 01:34:17] Received response 'proof/goals - (7)' in 27ms.
[Trace - 01:34:17] [process_queue]: resuming document checking
[Trace - 01:34:17] [check]: resuming [v: 40], from: 1125 l: 66
[Trace - 01:34:17] [resume]: last node :({ l: 64, c: 23 | o: 1102 }--{ l: 66, c: 1 | o: 1125 })
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] [recovery]: ({ l: 55, c: 0 | o: 940 }--{ l: 55, c: 30 | o: 970 }) stm: 547262707 | st 689204304
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] [recovery]: ({ l: 76, c: 0 | o: 1263 }--{ l: 76, c: 3 | o: 1285 }) stm: 547262707 | st 689204304
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] Received notification '$/coq/fileProgress'.
[Trace - 01:34:17] [check]: done [1674520457.44]: document fully checked with pos ({ l: 104, c: 0 | o: 1677 }--
{ l: 104, c: 0 | o: 1678 })
[Trace - 01:34:17] [cache]: hashing: 0.001452 | parsing: 0.022738 | exec: 1.138710
[Trace - 01:34:17] [cache]: cache hit rate: 6.67
[Trace - 01:34:17] Received notification 'textDocument/publishDiagnostics'.
[Trace - 01:34:17] Sending request 'textDocument/documentSymbol - (8)'.
[Trace - 01:34:17] [process_queue]: Serving Request: textDocument/documentSymbol
[Trace - 01:34:17] Received response 'textDocument/documentSymbol - (8)' in 20ms.
```

**Comment by ejgallego** [comment]
Ummm, thanks; that's very bizarre, just to be sure, can you reproduce with 48427e5db53afac5b7fbd43693265425d6dfb258 ?

If so, what do you have in that file, can you reproduce that with a small minimal file?

**Comment by Alizter** [comment]
I was in ex0.v and just made a new line and pressed `\`.

**Comment by ejgallego** [comment]
> I was in ex0.v and just made a new line and pressed `\`. Line 67 if that's important.

Works fine here :S

**Comment by Alizter** [comment]
@ejgallego Shall we do a quick meeting to debug together?

**Comment by artagnon** [comment]
Works for me too!

**Comment by artagnon** [comment]
@ejgallego I've added [completion_symbols.json](https://github.com/ejgallego/coq-lsp/pull/219/commits/e43da8746234a8afc26d1c2473affcd1e7c75821) that you should be able to wire up and provide a comprehensive list.

**Comment by Alizter** [comment]
@artagnon I was going to grab the existing table here: https://github.com/coq/coq/blob/master/ide/coqide/default_bindings_src.ml Maybe you can write a script that converts it to json?

**Comment by Alizter** [comment]
For some reason, when a line ends in `\` (which is pretty common), the last `\` character is dropped.

**Comment by Alizter** [comment]
Ah! Completion is being handled before didChange. This means that the doc it is operating on is stale. We should change that to be processed after didChange.

**Comment by Alizter** [comment]
Another interesting thing to note. If you do `\` after some constants symbols have been defined, i.e. with some definition names preceding, the completion does trigger, but it only shows these symbols. This is probably because the trigger character is not in the stale document.

**Comment by ejgallego** [comment]
@Alizter this is ready I think; can your trigger a problem with this version of the PR?

I'd be puzzled if that happens, if I understand my own logic if the position is invalid we should detect it by this check in `byte_of_char`:
```
  if ll <= char then ll else nth line char
```

**Comment by ejgallego** [comment]
By problem I mean a backtrace in the server.

**Comment by Alizter** [comment]
@ejgallego I still get the exception. But I think this is worth merging as is.

**Pull request merged**